bee-message/src/payload/transaction/essence.rs

```diff
@@ -0,0 +1,381 @@
+        self.access_pledge_id = Some(access_pledge_id);
+        self
+    }
+
+    /// Adds a consensus pledge ID to a `TransactionEssenceBuilder`.
+    pub fn with_consensus_pledge_id(mut self, consensus_pledge_id: [u8; PLEDGE_ID_LENGTH]) -> Self {
+        self.consensus_pledge_id = Some(consensus_pledge_id);
+        self
+    }
+
+    /// Adds inputs to a `TransactionEssenceBuilder`
+    pub fn with_inputs(mut self, inputs: Vec<Input>) -> Self {
+        self.inputs = inputs;
+        self
+    }
+
+    /// Add an input to a `TransactionEssenceBuilder`.
+    pub fn add_input(mut self, input: Input) -> Self {
+        self.inputs.push(input);
+        self
+    }
+
+    /// Add outputs to a `TransactionEssenceBuilder`.
+    pub fn with_outputs(mut self, outputs: Vec<Output>) -> Self {
+        self.outputs = outputs;
+        self
+    }
+
+    /// Add an output to a `TransactionEssenceBuilder`.
+    pub fn add_output(mut self, output: Output) -> Self {
+        self.outputs.push(output);
+        self
+    }
+
+    /// Add a payload to a `TransactionEssenceBuilder`.
+    pub fn with_payload(mut self, payload: Payload) -> Self {
+        self.payload = Some(payload);
+        self
+    }
+
+    /// Finishes a `TransactionEssenceBuilder` into a `TransactionEssence`.
+    pub fn finish(self) -> Result<TransactionEssence, ValidationError> {
+        let version = self.version.ok_or(ValidationError::MissingField("version"))?;
+        let timestamp = self.timestamp.ok_or(ValidationError::MissingField("timestamp"))?;
+        let access_pledge_id = self
+            .access_pledge_id
+            .ok_or(ValidationError::MissingField("access_pledge_id"))?;
+        let consensus_pledge_id = self
+            .consensus_pledge_id
+            .ok_or(ValidationError::MissingField("consensus_pledge_id"))?;
+
+        if !INPUT_OUTPUT_COUNT_RANGE.contains(&self.inputs.len()) {
+            return Err(ValidationError::InvalidInputCount(self.inputs.len()));
+        }
+
+        if !INPUT_OUTPUT_COUNT_RANGE.contains(&self.outputs.len()) {
+            return Err(ValidationError::InvalidOutputCount(self.outputs.len()));
+        }
+
+        if !matches!(self.payload, None | Some(Payload::Indexation(_))) {
+            // Unwrap is fine because we just checked that the Option is not None.
+            return Err(ValidationError::InvalidPayloadKind(self.payload.unwrap().kind()));
+        }
+
+        for input in self.inputs.iter() {
+            match input {
+                Input::Utxo(u) => {
+                    if self.inputs.iter().filter(|i| *i == input).count() > 1 {
+                        return Err(ValidationError::DuplicateUtxo(u.clone()));
+                    }
+                }
+            }
+        }
+
+        // Inputs must be lexicographically sorted in their serialised forms.
+        if !is_sorted(self.inputs.iter().map(Packable::pack_to_vec)) {
```

Comment:
Can't you take advantage of this and just check duplicates by checking that for every input, the next one is not the same ?

Reply:
That would work, but if there are ever any different `Input` variants added that could have duplicates and still be valid, it wouldn't. I'm not sure if this will ever be the case though...

Reply:
yeah I'd left it as it is then.